packages/contracts-bedrock/src/universal/UsdcBridge.sol

```diff
@@ -0,0 +1,362 @@
+    ///         UsdcBridge contract on the remote chain.
+    /// @param _localToken  Address of the ERC20 on this chain.
+    /// @param _remoteToken Address of the corresponding token on the remote chain.
+    /// @param _from        Address of the sender.
+    /// @param _to          Address of the receiver.
+    /// @param _amount      Amount of the ERC20 being bridged.
+    /// @param _extraData   Extra data to be sent with the transaction. Note that the recipient will
+    ///                     not be triggered with this data, but it will be emitted and can be used
+    ///                     to identify the transaction.
+    function finalizeBridgeERC20(
+        address _localToken,
+        address _remoteToken,
+        address _from,
+        address _to,
+        uint256 _amount,
+        bytes calldata _extraData
+    )
+        public
+        onlyOtherBridge
+    {
+        if (_isL2Usdc(_localToken) && _isL1Usdc(_remoteToken)) {
+            // L1 --> L2
+            IPartialUsdc(_localToken).mint(_to, _amount);
+        } else if (_isL1Usdc(_localToken) && _isL2Usdc(_remoteToken)) {
+            // L2 --> L1
+            deposits[_localToken][_remoteToken] = deposits[_localToken][_remoteToken] - _amount;
+            IERC20(_localToken).safeTransfer(_to, _amount);
+        } else {
+            revert("Invalid token pair");
+        }
+
+        // Emit the correct events. By default this will be ERC20BridgeFinalized, but child
+        // contracts may override this function in order to emit legacy events as well.
+        _emitERC20BridgeFinalized(_localToken, _remoteToken, _from, _to, _amount, _extraData);
+    }
+
+    /// @notice Sends ERC20 tokens to a receiver's address on the other chain.
+    /// @param _localToken  Address of the ERC20 on this chain.
+    /// @param _remoteToken Address of the corresponding token on the remote chain.
+    /// @param _to          Address of the receiver.
+    /// @param _amount      Amount of local tokens to deposit.
+    /// @param _minGasLimit Minimum amount of gas that the bridge can be relayed with.
+    /// @param _extraData   Extra data to be sent with the transaction. Note that the recipient will
+    ///                     not be triggered with this data, but it will be emitted and can be used
+    ///                     to identify the transaction.
+    function _initiateBridgeERC20(
+        address _localToken,
+        address _remoteToken,
+        address _from,
+        address _to,
+        uint256 _amount,
+        uint32 _minGasLimit,
+        bytes memory _extraData
+    )
+        internal
+        whenNotPaused
+    {
+        if (_isL2Usdc(_localToken) && _isL1Usdc(_remoteToken)) {
+            // L2 --> L1
+            // usdc has no burnFrom, so first transfer to the contract and then burn.
+            IERC20(_localToken).safeTransferFrom(_from, address(this), _amount);
+            IPartialUsdc(_localToken).burn(_amount);
```

Comment:
> Just for my understanding this is executed on the L2? 

yes

> And just to check _localToken is Circle's FiatToken?

yes

> That can only by burned by the owner right?

Can only be burned by addresses that have been assigned the `minter` role, as is the case for the bridge.

> Which is this contract? 

https://github.com/circlefin/stablecoin-evm/blob/657375471bff72afa5f625083bbab8003eb5f8c9/contracts/v1/FiatTokenV1.sol#L360

Reply:
Hmmm, I don't think the code is correct.

the function starts with:
require(_isCorrectUsdcTokenPair(_localToken, _remoteToken), "Invalid token pair");
and if you look into _isCorrectUsdcTokenPair for L2UsdcBridge

```
function _isCorrectUsdcTokenPair(address _localToken, address _remoteToken) internal view override returns (bool) {
        return _isL2Usdc(_localToken) && _isL1Usdc(_remoteToken);
    }

```
So if the require doesn't fail, then at this point I know that 
`  _isL2Usdc(_localToken) && _isL1Usdc(_remoteToken)`
  is true, so why do we have this if statement?
  `(_isL1Usdc(_localToken) && _isL2Usdc(_remoteToken))` when it will always be false?
  
  --------------
  Let's just think about this USDCBridge part of the L2UsdcBridge. If someone calls bridgeERC20To in our case, then someone want's to initiate an L2->L1 transfer. If someone initiated an L1->L2 transfer, then finalizeBridgeERC20 would have been called on the L2UsdcBridge and not the bridgeERC20To function.
  
  I think similar problem exists with the code in the finalizeBridgeERC20 method, but haven't checked.
  --------------
  The if/else statement in the original Optimum code is used to differentiate between different token implementations I think, for example:
  `_initiateBridgeERC20` case
  
  both branches initiate an L2->L1 transfer, but the first one works with 'burneable' tokens while the other token implementation might not be able to burn. For this second implementation we just transfer the local token from the user to the bridge and do a little bit of accounting in the deposit array to know how much of that non-burneable token is still there.
  
  if (_isOptimismMintableERC20(_localToken)) {
            require(
                _isCorrectTokenPair(_localToken, _remoteToken),
                "StandardBridge: wrong remote token for Optimism Mintable ERC20 local token"
            );

            OptimismMintableERC20(_localToken).burn(_from, _amount);
        } else {
            IERC20(_localToken).safeTransferFrom(_from, address(this), _amount);
            deposits[_localToken][_remoteToken] = deposits[_localToken][_remoteToken] + _amount;
        }